rfd/0183-access-request-kube-resource-allow-list.md

```diff
@@ -0,0 +1,225 @@
+
+The reviewer may not remember what this `kube_cluster` has access to and will have to look up the requested role to double check access which can be annoying.
+
+If a request came in for a subresource `namespace` for a `kube_cluster` instead, the reviewer sees this:
+
+| Cluster Name     | Requested Resource Kind |      Requested Resource Name |
+| ---------------- | :---------------------: | ---------------------------: |
+| teleport-cluster |        namespace        |     pumpkin-kube-cluster/dev |
+| teleport-cluster |        namespace        | pumpkin-kube-cluster/staging |
+
+The reviewer is more likely to understand what access is being granted.
+
+This RFD proposes a new request option that will enforce users to request for Kubernetes subresources instead.
+
+# User Scenarios
+
+Listing use cases from most permissive to most restrictive:
+
+## As an admin, I don't care what kind of kubernetes resources a user requests
+
+As a requester, this means I can request for any supported kubernetes resources (`kube_cluster` and its subresources).
+
+This will be the default behavior if no request options are specified. This will also be the default behavior of existing roles unless modified.
+
+## As an admin, I want to limit what namespaces a user can request
+
+This is already supported. The admin can use `kubernetes_resources` field to restrict what list of namespaces the requester can list and request for:
+
+```yaml
+kind: role
+metadata:
+  name: kube-access-request
+spec:
+  allow:
+    # Only lists namespaces starting with "pumpkin-" prefixes
+    kubernetes_resources:
+      - kind: namespace
+        name: pumpkin-*
+        verbs:
+          - list
+    kubernetes_groups:
+      - '{{internal.kubernetes_groups}}'
+    kubernetes_labels:
+      '*': '*'
+    kubernetes_users:
+      - '{{internal.kubernetes_users}}'
+```
+
+Requesting for namespace not in this list is denied.
+
+## As an admin, I want to limit Kubernetes resource requesting to pods and namespaces only
+
+Currently, users can request any of [these supported Kubernetes subresources](https://github.com/gravitational/teleport/blob/110b23aefb3c4b91c9d8cca594041b93f0e078dd/api/types/constants.go#L1233).
+
+Admins can specify in a role, a request Kubernetes allow list that limits request to just `pod` and `namespace` kind:
+
+```yaml
+kind: role
+metadata:
+  name: requester
+spec:
+  allow:
+    request:
+      # Only allows requesting for kube_cluster's pod and namespace resource
+      kubernetes:
+        allow: ['pod', 'namespace']
+```
+
+User requesting any other Kubernetes kind will get denied.
+
+## As an admin, I want to require users to only request namespaces
+
+Admins can specify in a role, a request Kubernetes allow list that only defines `namespace`.
+
+```yaml
+kind: role
+metadata:
+  name: requester
+spec:
+  allow:
+    request:
+      # Only allows requesting for kube_cluster's namespace resources
+      kubernetes:
+        allow: ['namespace']
```

Comment:
Putting this in the requester role still feels a bit off tbh, for a couple of reasons:

- The configuration for what you can actually request is now split between your requester role and the role you're requesting.
- This makes it harder to support scenarios like allowing requesters to request specific namespaces for one cluster, but everything (or another set of resources) for another cluster; you'd need to define multiple requester roles for that.

@kiosion is actually working on a somewhat similar feature request - making access request reason mandatory (https://github.com/gravitational/teleport/issues/20164), and we were considering making it an option on the _role being requested_ rather than requester role. Then you can have different requirements for resources granted by different roles.

Have you considered something similar here? Feels these two things are related and we should have a consistent approach. E.g. just off the top of my head if e.g. you have a role like this:

```yaml
spec:
  allow:
    kubernetes_resources:
      - kind: namespace
        name: pumpkin-*
        required: true
```

Then when someone is requesting the resource granted by this role, they have to select a namespace.

Reply:
My initial thought is that, since this feature is tied to Access Requests, we should clearly organize it under an "access request" block.

From my experience, customers often use roles as base roles for sysadmins or highly privileged users. This means that while some users might request access to role X, other users may be permanent members of that role.

With that in mind, the following configuration could be confusing:

```yaml
    kubernetes_resources:
      - kind: namespace
        name: pumpkin-*
        required: true
```

What does it mean for a permanent member of the role to have `required: true` in this context? I think this particular setting could be difficult to interpret.

I don’t dislike the idea of moving towards options for target roles, similar to this:

```yaml
kind: role
metadata:
  name: kube-access-request
spec:
  allow:
    request:
      search_as_roles:
      - devs
---
kind: role
version: v7
metadata:
  name: devs
spec:
  allow:
    logins: ['{{internal.logins}}']
    kubernetes_groups: ['{{internal.kubernetes_groups}}']
    node_labels:
      '*': '*'
    kubernetes_labels:
      '*': '*'
    kubernetes_resources:
      - kind: '*'
        name: '*'
        namespace: '*'
        verbs: ['*']
  options:
    access_request:
        reason: 'mandatory|optional'
        sub_resource: 'mandatory|optional'  # sub_resource is more generic and applies to namespaces, pods, secrets...
```

However, we need to be clear on the expected behavior when a user can request access with multiple roles that have conflicting options. Some of our options are "greedy" (where the most restrictive wins), while others only apply if the role is being used. I think we should define the expected behavior and user experience clearly in this case.


---

Placing it under the requestor role simplifies the behavior, as it provides a clear view of which roles can be requested and the rules governing the request, without needing to switch between different roles to assess the rules. For TAG, it’s also more straightforward to define the rules in the same location where we configure the `request` settings, similar to:

```yaml
kind: role
metadata:
  name: kube-access-request
spec:
  allow:
    request:
      options:
        reason: 'mandatory|optional'
        sub_resource: 'mandatory|optional'  # sub_resource is more generic and applies to namespaces, pods, secrets...
      search_as_roles:
      - devs
---
kind: role
version: v7
metadata:
  name: devs
spec:
  allow:
    logins: ['{{internal.logins}}']
    kubernetes_groups: ['{{internal.kubernetes_groups}}']
    node_labels:
      '*': '*'
    kubernetes_labels:
      '*': '*'
    kubernetes_resources:
      - kind: '*'
        name: '*'
        namespace: '*'
        verbs: ['*']
```

I am ok with any of the options above.

Reply:
going back to roman's suggestion, can we do something like this:

```
>> This means user can request either kube_cluster or its subresources   (default)
 kubernetes_resources:
    - kind: '*'                   ←—
      name: '*'
      namespace: '*'
      verbs:
      - '*'

>> This means user must specify a namespace:
kubernetes_resources:
      - kind: 'namespace'   ←– 
        name: '*'
        verbs: ['*']

```

so basically, if we find that `kind != *`, then we will disable requesting for the whole cluster?

Reply:
Not exactly, because they may have a kind other than '*' and still need access to the entire cluster. For example, someone might need access to nodes, and certificate signing requests (which are global resources). You wouldn't be able to replicate this access pattern since we won’t initially support all of those resource types

Reply:
i see.

i also think the main issue is the ability to `opt-in` to this feature (which won't disrupt current flows, and a admin might want to limit namespaces but allow whole cluster requests anyways). that means we _have_ to add a new field somewhere right?

if we go with roman's suggestion:

```
    kubernetes_resources:
      - kind: namespace
        name: pumpkin-*
        access_request_require: true     <--- opt-in
```

but if we are going to add a new field anyways, why not add a field in a central place? (whereever we end up putting it)

b/c i don't think it makes sense to do something like this:

```yaml
    kubernetes_resources:
     # restrict secrets
      - kind: secret
        access_request_require: true     # <--- disables whole cluster request
     # restrict pods
      - kind: pod
        access_request_require: false / true    # <--- isn't this redundant? or what does it mean if this means true as well?
```

Reply:
@kimlisa @tigrato To duplicate what we were discussing in Slack.

I was looking through our access requests documentation and configuration today and found some things that I didn't even know (or forgot) we support. Specifically, there are some additional request options that can be set on the requester role, for example: https://goteleport.com/docs/admin-guides/access-controls/access-requests/access-request-configuration/#how-clients-request-access. You can actually already specify things like "require request reason". It doesn't really work the way we need it to work for the thing @kiosion is working on (current impl. is tied to automatic access requests) but the point is that we already have something called [RequestAccess strategy](https://github.com/gravitational/teleport/blob/282d6d7ff5b1b8fc26da3d1c809685bcb75ea2fd/api/proto/teleport/legacy/types/types.proto#L2887-L2890), and a bunch of other options tied to the requester's role.

So with that in mind: I still find it more limiting than setting options on the roles being requested tbh but in order to not break any existing concepts and keep things consistent, I think I'm going to agree with the approach you both proposed initially and keeping this option as a part of the requester role. The only thing I think we still need to iron out is how to make the configuration more robust than just having a single `force_subresource_request` flag (which is a little confusing tbh) or even having `allow: ['namespace']` because that still doesn't allow any templating. So maybe we could do something like this:

```yaml
spec:
  allow:
    request:
      roles:
        - "dba"
      kubernetes:
        namespaces:
        - kube-system
        - {{external.kube_namespaces}}
        - *
```

Basically, if the namespaces option is present among the user's allow.request setting, we will force them to select namespaces when submitting an access request. If you, say, have a `*` there, you will be able to select any namespaces allowed by the role granting this cluster (assuming it also has `kube_resources` on it). Otherwise, you're limited to picking namespaces from the list of allowed ones in your request settings, which can also be templated from IdP attributes, like most other fields.

WDYGT?

Reply:
summarizing our slack discussion here:

we should keep the ability to define limits to kube subresources with the `kubernetes_resources` field since it now also supports trait interpolation (coming in v17 release).

agreed on adding a role request option using enum values `cluster | resource | both`

Reply:
updated RFD: https://github.com/gravitational/teleport/pull/46691/commits/26d187ff8611456d6b21b3ba9f28a71261d60df8